ledger-core/virtual/object/object_publish_callsummary_test.go

```diff
@@ -0,0 +1,149 @@
+	"github.com/stretchr/testify/require"
+	"testing"
+)
+
+func TestSMObject_CallSummarySM(t *testing.T) {
+	var (
+		mc = minimock.NewController(t)
+
+		smObject        = newSMObjectWithPulse()
+		sharedStateData = smachine.NewUnboundSharedData(&smObject.SharedState)
+	)
+
+	smObject.SetState(HasState)
+
+	res1 := payload.VCallResult{
+		Callee:       smObject.Reference,
+		CallOutgoing: gen.UniqueLocalRef(),
+	}
+	res2 := payload.VCallResult{
+		Callee:       smObject.Reference,
+		CallOutgoing: gen.UniqueLocalRef(),
+	}
+
+	req1Ref := reference.NewRecordOf(res1.Callee, res1.CallOutgoing)
+	req2Ref := reference.NewRecordOf(res2.Callee, res2.CallOutgoing)
+
+	smObject.SharedState.KnownRequests.GetList(contract.CallTolerable).Add(req1Ref)
+	smObject.SharedState.KnownRequests.GetList(contract.CallIntolerable).Add(req2Ref)
+	smObject.SharedState.KnownRequests.GetList(contract.CallTolerable).Finish(req1Ref, &res1)
+	smObject.SharedState.KnownRequests.GetList(contract.CallIntolerable).Finish(req2Ref, &res2)
+
+	pulseNumber := smObject.pulseSlot.PulseData().PulseNumber
+
+	{
+		initCtx := smachine.NewInitializationContextMock(mc).
+			ShareMock.Return(sharedStateData).
+			PublishMock.Expect(smObject.Reference.String(), sharedStateData).Return(true).
+			JumpMock.Return(smachine.StateUpdate{}).
+			SetDefaultMigrationMock.Set(func(m1 smachine.MigrateFunc) {
+
+		})
+
+		smObject.Init(initCtx)
+
+		_, inactive := smObject.SummaryDone.GetCounts()
+		require.Equal(t, 0, inactive)
+		limit, _ := smObject.SummaryDone.GetLimit()
+		require.Equal(t, 0, limit)
+	}
+
+	{
+		sdlStateReport := smachine.NewUnboundSharedData(&payload.VStateReport{})
+		sdlSummarySync := smachine.NewUnboundSharedData(&smachine.SyncLink{})
+
+		migrationCtx := smachine.NewMigrationContextMock(mc).
+			LogMock.Return(smachine.Logger{}).
+			UnpublishAllMock.Expect().Return().
+			ShareMock.Set(func(data interface{}, flags smachine.ShareDataFlags) (s1 smachine.SharedDataLink) {
+			switch data.(type) {
+			case *payload.VStateReport:
+				return sdlStateReport
+			case *smachine.SyncLink:
+				return sdlSummarySync
+			default:
+			}
+			panic("unexpected shared data")
+		}).
+			PublishMock.Set(func(key interface{}, data interface{}) (b1 bool) {
+			switch d := key.(type) {
+			case finalizedstate.ReportKey:
+				require.Equal(t, finalizedstate.BuildReportKey(smObject.Reference), d)
+			case callsummary.SummarySyncKey:
+				require.Equal(t, callsummary.BuildSummarySyncKey(smObject.Reference), d)
+			}
+			return true
+		}).
+			JumpMock.Set(testutils.AssertJumpStep(t, smObject.stepPublishCallSummary))
+
+		smObject.migrate(migrationCtx)
+	}
+
+	{
+		scs := callsummary.SharedCallSummary{
+			Requests: tables.NewObjectRequestTable(),
+		}
+		sdl := smachine.NewUnboundSharedData(&scs)
+
+		firsCall := true
+
+		executionCtx := smachine.NewExecutionContextMock(mc).
+			GetPublishedLinkMock.Set(func(key interface{}) (s1 smachine.SharedDataLink) {
```

Comment:
please, rewrite that function with something like:
```
var rv smachine.SharedDataLink
switch key {
case callsummary.BuildSummarySharedKey(pulseNumber):
    if !firstCall {
       firstCall = false
    } else {
        rv = sdl
    }
default:
    t.Fatal(...)
}
```

Reply:
and extract it before executionCtx creation

Reply:
Fixed